epochStart/bootstrap/storageProcess.go

```diff
@@ -0,0 +1,501 @@
+	cancel()
+	if err != nil {
+		return nil, err
+	}
+
+	syncedHeaders, err := sesb.headersSyncer.GetHeaders()
+	if err != nil {
+		return nil, err
+	}
+
+	if meta.Epoch == sesb.startEpoch+1 {
+		syncedHeaders[string(meta.EpochStart.Economics.PrevEpochStartHash)] = &block.MetaBlock{}
+	}
+
+	return syncedHeaders, nil
+}
+
+func (sesb *storageEpochStartBootstrap) processNodesConfig(pubKey []byte) error {
+	var err error
+	shardId := sesb.destinationShardAsObserver
+	if shardId > sesb.baseData.numberOfShards && shardId != core.MetachainShardId {
+		shardId = sesb.genesisShardCoordinator.SelfId()
+	}
+	argsNewValidatorStatusSyncers := ArgsNewSyncValidatorStatus{
+		DataPool:           sesb.dataPool,
+		Marshalizer:        sesb.coreComponentsHolder.InternalMarshalizer(),
+		RequestHandler:     sesb.requestHandler,
+		ChanceComputer:     sesb.rater,
+		GenesisNodesConfig: sesb.genesisNodesConfig,
+		NodeShuffler:       sesb.nodeShuffler,
+		Hasher:             sesb.coreComponentsHolder.Hasher(),
+		PubKey:             pubKey,
+		ShardIdAsObserver:  shardId,
+	}
+	sesb.nodesConfigHandler, err = NewSyncValidatorStatus(argsNewValidatorStatusSyncers)
+	if err != nil {
+		return err
+	}
+
+	clonedHeader := sesb.epochStartMeta.Clone()
+	clonedEpochStartMeta, ok := clonedHeader.(*block.MetaBlock)
+	if !ok {
+		return fmt.Errorf("%w while trying to assert clonedHeader to *block.MetaBlock", epochStart.ErrWrongTypeAssertion)
+	}
+	sesb.applyCurrentShardIDOnMiniblocksCopy(clonedEpochStartMeta)
+
+	clonedHeader = sesb.prevEpochStartMeta.Clone()
+	clonedPrevEpochStartMeta, ok := clonedHeader.(*block.MetaBlock)
+	if !ok {
+		return fmt.Errorf("%w while trying to assert prevClonedHeader to *block.MetaBlock", epochStart.ErrWrongTypeAssertion)
+	}
+	sesb.applyCurrentShardIDOnMiniblocksCopy(clonedPrevEpochStartMeta)
+
+	sesb.nodesConfig, sesb.baseData.shardId, err = sesb.nodesConfigHandler.NodesConfigFromMetaBlock(clonedEpochStartMeta, clonedPrevEpochStartMeta)
+	sesb.baseData.shardId = sesb.applyShardIDAsObserverIfNeeded(sesb.baseData.shardId)
+
+	return err
+}
+
+// applyCurrentShardIDOnMiniblocksCopy will alter the fetched metablocks making the sender shard ID for each miniblock
+// header to  be exactly the shard ID used in the import-db process. This is necessary as to allow the miniblocks to be requested
+// on the available resolver and should be called only from this storage-base bootstrap instance.
+// This method also copies the MiniBlockHeaders slice pointer. Otherwise the node will end up stating
+// "start of epoch metablock mismatch"
+func (sesb *storageEpochStartBootstrap) applyCurrentShardIDOnMiniblocksCopy(metablock *block.MetaBlock) {
+	originalMiniblocksHeaders := metablock.MiniBlockHeaders
+	metablock.MiniBlockHeaders = make([]block.MiniBlockHeader, 0, len(originalMiniblocksHeaders))
+	for _, mbh := range originalMiniblocksHeaders {
+		mbh.SenderShardID = sesb.importDbConfig.ImportDBTargetShardID //it is safe to modify here as mbh is passed by value
```

Comment:
The actual mbh is changed, is it really passed by value?

Reply:
yup, as seen in this test, please try it and then alter `original := []testStruct` to `original := []*testStruct`:
```GO
type testStruct struct{
    data string
}

func TestCopyForRange(t *testing.T){
    original := []testStruct{
        {
            data: "data1",
        },
        {
            data: "data2",
        },
    }

    for _, ts := range original{
        ts.data = "altered"
    }

    for _, ts := range original{
        fmt.Println(ts.data)
    }
}
```

Reply:
yes... range does a copy